File: CSharp/CSFundamentals/DataStructures/Stack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CSFundamentals.DataStructures
{
    public class Stack<T>
    {
        private SinglyLinkedList<T> stack = new SinglyLinkedList<T>();

        /// <summary>
        /// Pushes the value to the top of the stack.
        /// </summary>
        /// <param name="value"></param>
        public void Push(T value)
        {
            stack.AddToBeginning(value);
        }

        /// <summary>
        /// Returns the top value of the stack without removing the item.
        /// </summary>
        /// <returns></returns>
        public T Peek()
        {
            if(stack.Head == null)
            {
                return default(T);
            }

            return stack.Head.Value;
        }

        /// <summary>
        /// Returns and removes the top value of the stack.
        /// </summary>
        /// <returns></returns>
        public T Pop()
        {
            T value = stack.GetAndRemoveHead();
            return value;
        }
    }
}

File: Java/src/Stack.java

public class Stack<T extends Comparable<T>> {
	private LinkedList<T> list = new LinkedList<T>();
	
	private int size = 0;
	
	public int getSize() {
		return size;
	}

	/**
	 * Returns the top of the stack. Returns null if the stack is empty.
	 */
	public T peek() {
		if(list == null) return null;
		if(list.getHeadValue() == null) return null;
		return list.getHeadValue();
	}

	/**
	 * Pushes the value to the top of the stack. Does nothing if the value is null.
	 */
	public void push(T value) {
		if(value == null) return;
		list.addToHead(value);
		size++;
	}

	/**
	 * Returns the top of the stack and removes it from the stack. Returns null if the stack is empty.
	 */
	public T pop() {
		if(list == null) return null;
		if(list.getHeadValue() == null) return null;
		T topValue = list.getHeadValue();
		list.removeHead();
		size--;
		return topValue;
	}
}
